### ISSUE TITLE
Reliable Topic not working as expected 

### ISSUE THREAD

**Issue by dmartinsid** [hazelcast/hazelcast#19327]
**Describe the bug**
When a Hazelcast Server publish a message to a reliable topic and this reliable topic have clients in Python and Java occurs the follow bug. 
The Python Hazelcast client don't receive the message, but the Java client receive it.
This behavior not occurs to a standard Topic

This bug not occurs if a client Python or client Java publish to that reliable topic, in this scenario all the clients receive the desire message.

**Expected behavior**
The expected behavior is when we publish to a reliable topic using a Hazelcast Server this message be distributed to all clients in all platforms not only Java

**To Reproduce**

1. Start Java Server, files here are Server.java e it's pom.xml
2. Start Java Client, files here are Client.java e it's pom.xml
3. Start Python client, file here is client.py
4. See that in python logs never came a message from java server, only from java client e python client.
5. See that in java client logs we see messages from java server, java client e python client

See here some logs

**Python Client Logs, only events from Java Server from a standard topic**

```shell
Python Client Received: Client Python publishing to RELIABLE topic 2021-08-13T15:45:03.282164
Python Client Received: Client Python publishing to NORMAL topic2021-08-13T15:45:03.288156
Python Client Received: Server publishing to NORMAL topic 2021-08-13T15:45:07.767
Python Client Received: Client Java publishing to NORMAL topic 2021-08-13T15:45:22.533
Python Client Received: Client Java publish to RELIABLE topic 2021-08-13T15:45:22.535
Python Client Received: Client Python publishing to RELIABLE topic 2021-08-13T15:45:33.294065
Python Client Received: Client Python publishing to NORMAL topic2021-08-13T15:45:33.297975
Python Client Received: Server publishing to NORMAL topic 2021-08-13T15:45:37.771
Python Client Received: Client Java publishing to NORMAL topic 2021-08-13T15:45:52.539
Python Client Received: Client Java publish to RELIABLE topic 2021-08-13T15:45:52.541
```

**Java Client Logs, all events from reliable and standard topic**

```shell
Java Client Received: Client Java publishing to NORMAL topic 2021-08-13T15:44:52.501
Java Client Received: Client Java publish to RELIABLE topic 2021-08-13T15:44:52.504
Java Client Received: Client Python publishing to RELIABLE topic 2021-08-13T15:45:03.282164
Java Client Received: Client Python publishing to NORMAL topic2021-08-13T15:45:03.288156
Java Client Received: Server publishing to NORMAL topic 2021-08-13T15:45:07.767
Java Client Received: Server publish to RELIABLE topic 2021-08-13T15:45:07.767
Java Client Received: Client Java publishing to NORMAL topic 2021-08-13T15:45:22.533
Java Client Received: Client Java publish to RELIABLE topic 2021-08-13T15:45:22.535
Java Client Received: Client Python publishing to RELIABLE topic 2021-08-13T15:45:33.294065
Java Client Received: Client Python publishing to NORMAL topic2021-08-13T15:45:33.297975
Java Client Received: Server publishing to NORMAL topic 2021-08-13T15:45:37.771
Java Client Received: Server publish to RELIABLE topic 2021-08-13T15:45:37.771
```
**Java Server Logs, all events from reliable and standard topic**

```
Server Received: Client Python publishing to RELIABLE topic 2021-08-13T15:45:03.282164
Server Received: Client Python publishing to NORMAL topic2021-08-13T15:45:03.288156
Server Received: Server publishing to NORMAL topic 2021-08-13T15:45:07.767
Server Received: Server publish to RELIABLE topic 2021-08-13T15:45:07.767
Server Received: Client Java publishing to NORMAL topic 2021-08-13T15:45:22.533
Server Received: Client Java publish to RELIABLE topic 2021-08-13T15:45:22.535
Server Received: Client Python publishing to RELIABLE topic 2021-08-13T15:45:33.294065
Server Received: Client Python publishing to NORMAL topic2021-08-13T15:45:33.297975
Server Received: Server publishing to NORMAL topic 2021-08-13T15:45:37.771
Server Received: Server publish to RELIABLE topic 2021-08-13T15:45:37.771
Server Received: Client Java publishing to NORMAL topic 2021-08-13T15:45:52.539
Server Received: Client Java publish to RELIABLE topic 2021-08-13T15:45:52.541
```
**All the files to reproduce bellow**

#### **`Server.java`**
```java
package com.hazelcast;

import java.text.MessageFormat;
import java.time.LocalDateTime;

import com.hazelcast.config.Config;
import com.hazelcast.config.ReliableTopicConfig;
import com.hazelcast.config.RingbufferConfig;
import com.hazelcast.core.Hazelcast;
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.topic.ITopic;
import com.hazelcast.topic.Message;
import com.hazelcast.topic.MessageListener;
import com.hazelcast.topic.TopicOverloadPolicy;

public class Server 
{
	private static final String RELIABLE_TOPIC_NAME = "reliable";
	private static final String NORMAL_TOPIC_NAME = "normal";

	public static void main(final String[] args) {
		
		final HazelcastInstance hz = Hazelcast.newHazelcastInstance(configureReliableTopic());
		
        final ITopic<String> reliable = hz.getReliableTopic(RELIABLE_TOPIC_NAME);
        final ITopic<String> normal = hz.getTopic(NORMAL_TOPIC_NAME);
        
        reliable.addMessageListener(new MessageListenerImpl());
        normal.addMessageListener(new MessageListenerImpl());
        
       
        while(true) {
        	publishMessagesEvery30seconds(reliable, normal);
        }
    }

	private static void publishMessagesEvery30seconds(final ITopic<String> reliable, final ITopic<String> normal) {
		normal.publish(MessageFormat.format("Server publishing to NORMAL topic {0}", LocalDateTime.now()));
		reliable.publish(MessageFormat.format("Server publish to RELIABLE topic {0}", LocalDateTime.now()));
		
		wait30Seconds();
	}


	private static Config configureReliableTopic() {
		final Config config = new Config();
		
		final RingbufferConfig ringbufferConfig = new RingbufferConfig(RELIABLE_TOPIC_NAME);
        ringbufferConfig.setCapacity(10000);
        config.addRingBufferConfig(ringbufferConfig);

        final ReliableTopicConfig topicConfig = new ReliableTopicConfig(RELIABLE_TOPIC_NAME);
        topicConfig
                .setTopicOverloadPolicy( TopicOverloadPolicy.BLOCK )
                .setReadBatchSize(10);
        
        return config;
	}
	
	private static void wait30Seconds() {
		try {
			Thread.sleep(30000);
		} catch (final InterruptedException e) {
		}
	}

    private static class MessageListenerImpl implements MessageListener<String> {
        public void onMessage(final Message<String> message) {
            System.out.println(MessageFormat.format("Server Received: {0}", message.getMessageObject()));
        }
    }

}
```
#### **`pom.xml`**
```xml
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>

  <groupId>com.hazelcast</groupId>
  <artifactId>reliable-topic-server</artifactId>
  <version>0.0.1-SNAPSHOT</version>
  <packaging>jar</packaging>

  <name>reliable-topic-server</name>
  <url>https://github.com/hazelcast</url>

  <properties>
    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    <java.version>1.8</java.version>
		<maven.compiler.source>1.8</maven.compiler.source>
		<maven.compiler.target>1.8</maven.compiler.target>
  </properties>

  <dependencies>
        <dependency>
            <groupId>com.hazelcast</groupId>
            <artifactId>hazelcast-all</artifactId>
            <version>4.2.2</version>
        </dependency>
  </dependencies>
</project>
```

#### **`Client.java`**
```java
package com.hazelcast;

import java.text.MessageFormat;
import java.time.LocalDateTime;

import com.hazelcast.client.HazelcastClient;
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.topic.ITopic;
import com.hazelcast.topic.Message;
import com.hazelcast.topic.MessageListener;

public class Client 
{
	private static final String RELIABLE_TOPIC_NAME = "reliable";
	private static final String NORMAL_TOPIC_NAME = "normal";

	public static void main(final String[] args) {
		
		final HazelcastInstance hz = HazelcastClient.newHazelcastClient();
		
        final ITopic<String> reliable = hz.getReliableTopic(RELIABLE_TOPIC_NAME);
        final ITopic<String> normal = hz.getTopic(NORMAL_TOPIC_NAME);
        
        reliable.addMessageListener(new MessageListenerImpl());
        normal.addMessageListener(new MessageListenerImpl());
        
       
        while(true) {
        	publishMessagesEvery30seconds(reliable, normal);
        }
    }

	private static void publishMessagesEvery30seconds(final ITopic<String> reliable, final ITopic<String> normal) {
		normal.publish(MessageFormat.format("Client Java publishing to NORMAL topic {0}", LocalDateTime.now()));
		reliable.publish(MessageFormat.format("Client Java publish to RELIABLE topic {0}", LocalDateTime.now()));
		
		wait30Seconds();
	}


	private static void wait30Seconds() {
		try {
			Thread.sleep(30000);
		} catch (final InterruptedException e) {
		}
	}

    private static class MessageListenerImpl implements MessageListener<String> {
        public void onMessage(final Message<String> message) {
            System.out.println(MessageFormat.format("Java Client Received: {0}", message.getMessageObject()));
        }
    }

}
```

#### **`pom.xml`**
```xml
<project xmlns="http://maven.apache.org/POM/4.0.0"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>

	<groupId>com.hazelcast</groupId>
	<artifactId>reliable-topic-client</artifactId>
	<version>0.0.1-SNAPSHOT</version>
	<packaging>jar</packaging>

	<name>reliable-topic-client</name>
	<url>http://maven.apache.org</url>

	<properties>
		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
		<java.version>1.8</java.version>
		<maven.compiler.source>1.8</maven.compiler.source>
		<maven.compiler.target>1.8</maven.compiler.target>
	</properties>

	<dependencies>
		<dependency>
			<groupId>com.hazelcast</groupId>
			<artifactId>hazelcast</artifactId>
			<version>4.2.2</version>
		</dependency>
	</dependencies>
</project>

```

#### **`client.py`**
```python
import hazelcast
import time
from datetime import datetime

client = hazelcast.HazelcastClient()

reliable_topic = client.get_reliable_topic("reliable").blocking()
normal_topic = client.get_topic("normal").blocking()

message_listener = lambda m: print("Python Client Received:", m.message)

reliable_topic.add_listener(message_listener)
normal_topic.add_listener(message_listener)

while True:
    reliable_topic.publish("Client Python publishing to RELIABLE topic " + datetime.now().isoformat())
    normal_topic.publish("Client Python publishing to NORMAL topic" + datetime.now().isoformat())

    time.sleep(30)


```

**Comment by dmartinsid** [comment]
* System MacOS
* openjdk version "1.8.0_252" - OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_252-b09) - OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.252-b09, mixed mode)
* Python 3.9.6 (default, Jun 29 2021, 06:20:32) - [Clang 12.0.0 (clang-1200.0.32.29)] on darwin

**Comment by mdumandag** [comment]
Hi @dmartinsid. Thanks a lot for the detailed bug report.

I was able to reproduce this and sent a fix https://github.com/hazelcast/hazelcast-python-client/pull/449, as this is a problem in the Python client-side.

Long story short, reliable topic messages are received as IdentifiedDataSerialiable objects and it has a field for the member address. Clients set this address to `null`, but members set it to a proper address.

Python client was not able to read Address objects written in IdentifiedDataSerialiable objects and that was the reason behind the failure. With the PR above, it will have this functionality and I have added a test that verifies the client is able to read messages written by a server.

We will include this fix in the 4.2.1 version of the Python client, which should be released approximately in a week. 

I am closing this issue now. Feel free to open up issues in the Python client if you have further questions.

Have a great day!
